server/datastore/mysql/migrations/tables/20240813103800_SupportMultipleABMTokens.go

```diff
@@ -0,0 +1,192 @@
+	var token []byte
+	if err := txx.Get(&token, getABM, fleet.MDMAssetABMTokenDeprecated); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// nothing to migrate, exit early
+			return nil
+		}
+		return fmt.Errorf("selecting team configs: %w", err)
+	}
+
+	// get the current ABM configuration from the app config
+	const getABMCfg = `
+SELECT
+	json_value->"$.mdm"
+FROM
+	app_config_json
+LIMIT 1
+`
+	var raw json.RawMessage
+	if err := txx.Get(&raw, getABMCfg); err != nil {
+		return fmt.Errorf("select MDM config from app_config_json: %w", err)
+	}
+
+	var (
+		abmTermsExpired bool
+		abmDefaultTeam  string
+	)
+	// decode if we did get an object
+	if len(raw) > 0 && raw[0] == '{' {
+		var config map[string]interface{}
+		if err := json.Unmarshal(raw, &config); err != nil {
+			return fmt.Errorf("unmarshal appconfig: %w", err)
+		}
+
+		if s, ok := config["apple_bm_default_team"].(string); ok {
+			abmDefaultTeam = s
+		}
+		if b, ok := config["apple_bm_terms_expired"].(bool); ok {
+			abmTermsExpired = b
+		}
+	}
+
+	var defaultTeamID *uint
+	if abmDefaultTeam != "" {
+		// get the default team id
+		var id uint
+		if err := txx.Get(&id, "SELECT id FROM teams WHERE name = ?", abmDefaultTeam); err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				return fmt.Errorf("select default ABM team id: %w", err)
+			}
+		}
+		// only use it if the team exists
+		if id > 0 {
+			defaultTeamID = &id
+		}
+	}
+
+	// NOTE: we don't know the organization_name, apple_id and renew_at of the
+	// existing token - the only way to know is to make an Apple API call and to
+	// decrypt the token, two operations that are probably not safe to do in a DB
+	// migration. Instead, insert with empty values and add a job at Fleet
+	// startup to ensure the token's information is filled.
+	// TODO(mna): e.g. insert a one-off worker job message?
+
+	// insert the token in the new table
+	const insABM = `
+INSERT INTO abm_tokens
+	(
+		organization_name,
+		apple_id,
+		terms_expired,
+		renew_at,
+		token,
+		macos_default_team_id,
+		ios_default_team_id,
+		ipados_default_team_id
+	)
+VALUES
+	('', '', ?, DATE('2000-01-01'), ?, ?, ?, ?)
+`
+	res, err := tx.Exec(insABM, abmTermsExpired, token, defaultTeamID, defaultTeamID, defaultTeamID)
+	if err != nil {
+		return fmt.Errorf("insert existing ABM token into abm_tokens: %w", err)
+	}
+	tokenID, _ := res.LastInsertId()
+
+	// soft-delete the token from the deprecated storage
+	const delABM = `
+UPDATE
+	mdm_config_assets
+SET
+	deleted_at = CURRENT_TIMESTAMP(),
+	deletion_uuid = ?
+WHERE
+	name = ? AND deletion_uuid = ''
+`
+	deletionUUID := uuid.New().String()
+	if _, err = tx.Exec(delABM, deletionUUID, fleet.MDMAssetABMTokenDeprecated); err != nil {
+		return fmt.Errorf("delete ABM token from mdm_config_assets: %w", err)
+	}
+
+	// associate all existing host DEP enrollments with the existing token
+	const updDEP = `
+UPDATE
+	host_dep_assignments
+SET
+	abm_token_id = ?
+WHERE
+	deleted_at IS NULL
```

Comment:
Does that look ok to update all non-soft-deleted rows in `host_dep_assignments`? I'm not super familiar with that table, but it looks like it stores whatever we sync from ABM and so all rows in there should be associated with the existing token?

Reply:
in ABM there are two calls to sync, one that retrieves all records, and another you poll that retrieves recent updates.

I don't remember if we're doing the first one periodically. If we aren't then we should try to avoid this, or do it once every time the server starts? (maybe we already do some variation of this)

Reply:
> If we aren't then we should try to avoid this

Sorry I'm not following, what should we avoid? The UPDATE above? Shouldn't all existing rows in that table be associated with the existing token?

Reply:
argh, I'm dumb, this is fine. Sorry!